### DISCUSSION TITLE
Upgrade of certain packages doesn't work

### DISCUSSION THREAD

**Discussion by toonetown** [Homebrew/discussions#14] [Everyday usage] [answered]
For example:
```
$ brew upgrade --cask dash
Warning: Treating dash as a formula. For the cask, use homebrew/cask/dash
```

If I'm specifically specifying `--cask` in the command line, shouldn't it treat it as a cask instead of a formula?  I see the same if you try to `brew upgrade --cask wireshark` (and probably any other cask that has a formula of the same name).

**Comment by SMillerDev** [comment, marked as answer]
No, those flags are only supposed to be used without specifying a target. It wouldn't make sense to support this since the target name already specifies if it's a cask or formula.

**Comment by toonetown** [comment]
> the target name already specifies if it's a cask or formula.

Except for in the case of a collision - like this - where it could be the name of either a cask _or_ a formula. 😉 

I would like to argue that `brew upgrade <caskname>` should be a bit smarter in the case of collisions.  If I _only_ have the wireshark cask installed, then `brew upgrade wireshark` should upgrade the installed cask (it is unambiguous, because you only upgrade what is installed).  

Only when I have _both_ a formula and a cask (or two formulas - or two casks - from additional taps) **installed** with the same short name should `brew upgrade wireshark` fail (and - I would argue it should fail, because if both are installed, the it is ambiguous as to what you want to upgrade - and falling back to cask is just as likely the desired behavior as falling back to formula).

I am willing to put together a PR if this would be helpful and avoid confusion - but I don't want to step on toes if this would not be a desired or welcome change in behavior.

**Reply by SMillerDev** [reply]
> Except for in the case of a collision - like this - where it could be the name of either a cask or a formula. 😉

It wouldn't since the cask name includes the `homebrew/cask/` part.

> I would like to argue that brew upgrade <caskname> should be a bit smarter in the case of collisions. If I only have the wireshark cask installed, then brew upgrade wireshark should upgrade the installed cask (it is unambiguous, because you only upgrade what is installed).

You could make a PR for this, but I'm not sure how other maintainers would feel about this.

**Reply by reitermarkus** [reply]
I think `brew upgrade --cask <cask>` should force the argument to be treated as a cask.

**Comment by toonetown** [comment]
> It wouldn't since the cask name includes the homebrew/cask/ part.

Then why isn't the same argument valid for including the `homebrew/brew/` part on formulas?

I'm planning on working around it for my use case by using `brew list --cask --full-name` and `brew list --full-name`.  So - I'm good.  I'll wait to hear on feedback from other maintainers before digging in to make a PR.

The main point I'm making is that there is inconsistency.  The recent changes to move `brew cask <action>` to just have `brew <action>` be "smart" enough to handle casks has been great for consistency, IMO - but I think that there might be some areas where the intelligence around it could be improved.

(Note: This is in no way meant to be a criticism...this is _your_ project, I'm just a visitor.  I'm just trying to get a read on some of the logic behind why things are done so that I can better help and support the community in the way it wants to be supported.  I don't want to be an outsider coming in and trying to change things for the sake of change...so your answer previous answers have been completely valid, and I accept them as-is.)
